feat(langchain): finishing up HITL implementation

sydney-runkle opened the pull request:
# Main changes / new features

## Better support for parallel tool calls

1. Support for multiple tool calls requiring human input
2. Support for combination of tool calls requiring human input + those that are auto-approved
3. Support structured output w/ tool calls requiring human input
4. Support structured output w/ standard tool calls

## Enhanced action request / response primitives

Simplified the primitives we use for action responses -- `accept` and `reject` patterns. `edit` is rolled into `accept` w/ edits.

These primitives have no notion of tool calls (no ids etc) so we can generically use them for HITL patterns independent of tool call approvals.

## Shortcut for allowed actions

Adds a shortcut where tool config can be specified as a `bool`, meaning "all actions allowed"

```py
HumanInTheLoopMiddleware(tool_configs={"expensive_tool": True})
```

## A few design decisions here
* We only raise one interrupt w/ all `HumanInterrupt`s, currently we won't be able to execute all tools until all of these are resolved. This isn't super blocking bc we can't re-invoke the model until all tools have finished execution. That being said, if you have a long running auto-approved tool, this could slow things down.
* We populate the `args` for an action request w/ `tool_name` and `tool_args` and expect the same for edits.
* `allowed_actions` is now represented as a list to be consistent with most other access control patterns

# TODO

* need to do a big docs pass
* in another PR I'd like to refactor testing to have one file for each prebuilt middleware :)
* work w/ applied AI team to ensure compat w/ agent inbox long term

sydney-runkle committed:
HITL extension

sydney-runkle committed:
linting

sydney-runkle committed:
remove helper function

sydney-runkle committed:
basic testing + fix w/ shadow var

sydney-runkle commented:
> If this works for now, it's probably fine? Behavior is correct as far as an end user is concerned. We could add a TODO for this refactor (and potentially apply the patch I made before for ToolNode).

Sounds great

sydney-runkle commented:
Pasting here, doesn't belong in description anymore:

Main question here is what we do with multiple tool calls and how we send them to tool node:
    If we use `Send`:
    * `interrupt`s within `tools` work as expected bc we kick off X `ToolNode`s in parallel where X is the number of pending tools
    * We have to inject tool args ahead of time which is quite messy, we could instead refactor to pass state through (@eyurtsev did this at some point)
    
    If we don't use `Send`
    * `interrupt`s within tools don't work as expected :(
    * We need to modify logic in `ToolNode` to filter through tool calls and only execute pending ones
    
    I think the value of interrupts working as expected is high, so perhaps should just refactor injection.

Will open an issue advocating for revisiting this

sydney-runkle committed:
adding new primitives, error handling for invalid response types, mapping tool call id -> resume

sydney-runkle committed:
remove comment

sydney-runkle committed:
test w/ new boolean logic

eyurtsev reviewed (commented):
Flash review, looking at the code

eyurtsev commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,268 @@
     def __init__(
         self,
-        tool_configs: ToolInterruptConfig,
+        tool_configs: dict[str, bool | HumanInterruptConfig],
```
What is the expected behavior for tool calls that haven't been configured via tool_configs? (Could we document what we expect the middleware to do? e.g., raise or auto-approve?)

eyurtsev commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,268 @@
+                * `True` indicates all interrupt config options are allowed.
+                * `False` indicates that the tool is auto-approved.
+                * `HumanInterruptConfig` indicates the specific interrupt config options to use.
             message_prefix: The message prefix to use when constructing interrupt content.
```
`message_prefix` is not also not self explanatory -- another sentence would help maybe?

eyurtsev commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,268 @@
         self,
-        tool_configs: ToolInterruptConfig,
+        tool_configs: dict[str, bool | HumanInterruptConfig],
         message_prefix: str = "Tool execution requires approval",
```
Could we `*` prior to message_prefix?

sydney-runkle commented on libs/langchain_v1/tests/unit_tests/agents/test_middleware_agent.py:
```diff
@@ -397,11 +399,11 @@ def test_human_in_the_loop_middleware_no_interrupts_needed() -> None:
     tool_configs = {
         "test_tool": HumanInterruptConfig(
-            allow_ignore=True, allow_respond=True, allow_edit=True, allow_accept=True
+            allow_approve=True, allow_ignore=True, allow_response=True, allow_edit=True
```
Yep, see https://github.com/langchain-ai/langchain/pull/32962#discussion_r2352752216

sydney-runkle commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,268 @@
+class ResponsePayload(TypedDict):
+    """Human provided text feedback or instructions."""
+
+    type: Literal["response"]
```
That's a good question, practically nothing at the moment given that `ignore` now allows for tool message customization. So maybe `ignore` should pop the tool call from the `AIMessage`?

eyurtsev replied:
Is `ignore` equivalent to `delete` ?

sydney-runkle commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,268 @@
+
+    type: Literal["ignore"]
+    tool_call_id: str
+    tool_message: NotRequired[str | ToolMessage]
```
Yep, using a default message right now.

sydney-runkle committed:
docs improvements

sydney-runkle commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,268 @@
+
+    Attributes:
+        allow_approve: Whether the human can accept/approve the current state
+        allow_ignore: Whether the human can choose to ignore/skip the current step
```
Added, though as noted in comments below, needs further refinement

sydney-runkle committed:
going back to sequence based approach

sydney-runkle committed:
feat(langchain): support for parallel (or interrupted) tool calls and structured output (#32980)

This enables parallel tool calling w/ a combo of
1. Standard and structured response tool calls
2. Deferred (requiring human approval / edits) tool calls and structured
response tool calls

Hard to unit test w/ HITL right now end to end, so here's a repro of
things working w/ an integration test:

```py
from pydantic import BaseModel, Field
from langchain_core.tools import tool
from langchain_core.messages import HumanMessage
from langgraph.types import Command
from langgraph.checkpoint.memory import InMemorySaver
from langchain.agents.middleware_agent import create_agent
from langchain.agents.middleware.human_in_the_loop import HumanInTheLoopMiddleware
from langchain_openai import ChatOpenAI


class WeatherBaseModel(BaseModel):
    temperature: float = Field(description="Temperature in fahrenheit")
    condition: str = Field(description="Weather condition")


@tool
def add_numbers(a: int, b: int) -> int:
    """Add two numbers."""
    return a + b


model = ChatOpenAI(model="gpt-4o-mini", temperature=0)
checkpointer = InMemorySaver()

agent = create_agent(
    model=model,
    tools=[add_numbers],
    response_format=WeatherBaseModel,
    middleware=[HumanInTheLoopMiddleware(tool_configs={"add_numbers": True})],
)
agent = agent.compile(checkpointer=checkpointer)

# First invocation should be interrupted due to human-in-the-loop middleware
response = agent.invoke(
    {
        "messages": [
            HumanMessage(
                "Add 1 and 2, then return the weather forecast with temperature 72 and condition sunny."
            )
        ]
    },
    config={"configurable": {"thread_id": "1"}},
)
interrupt_description = response["__interrupt__"][0].value[0]["description"]
print(interrupt_description)
"""
Tool execution requires approval

Tool: add_numbers
Args: {'a': 1, 'b': 2}
"""

# Resume the agent with approval
response = agent.invoke(
    Command(resume=[{"type": "approve"}]), config={"configurable": {"thread_id": "1"}}
)

for msg in response["messages"]:
    msg.pretty_print()

"""
================================ Human Message =================================

Add 1 and 2, then return the weather forecast with temperature 72 and condition sunny.
================================== Ai Message ==================================
Tool Calls:
  WeatherBaseModel (call_u6nXsEYRJbqNx4AEHgiQMpE2)
 Call ID: call_u6nXsEYRJbqNx4AEHgiQMpE2
  Args:
    temperature: 72
    condition: sunny
  add_numbers (call_nuQEZF7PwfYDlVpnSt8eaInI)
 Call ID: call_nuQEZF7PwfYDlVpnSt8eaInI
  Args:
    a: 1
    b: 2
================================= Tool Message =================================
Name: WeatherBaseModel

Returning structured response: temperature=72.0 condition='sunny'
================================= Tool Message =================================
Name: add_numbers

3
"""

print(repr(response["response"]))
"""
WeatherBaseModel(temperature=72.0, condition='sunny')
"""

```

sydney-runkle committed:
using new primitives

sydney-runkle commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,203 @@
-        message_prefix: str = "Tool execution requires approval",
+        tool_configs: dict[str, bool | list[AllowedActions]],
+        *,
+        action_request_prefix: str = "Tool execution requires approval",
```
I think this is more clear

sydney-runkle committed:
update tests

eyurtsev commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,201 @@
+class ApproveResponse(TypedDict):
+    """Response when a human approves the action."""
+
+    action: Literal["approve"]
```
name collision in `action`. `action` refers the action taken by the reviewer / HIL as well as to the action taken by the agent.

sydney-runkle committed:
Merge branch 'master' into sr/hitl-improvements

eyurtsev commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,201 @@
 
-ToolInterruptConfig = dict[str, HumanInterruptConfig]
+
+class ActionRequest(TypedDict):
```
I don't think `ActionRequest` works well in terms of communicating what the purpose is. `HumanInterrupt` actually made is much simpler to understand that something was being sent for review and involving interrupting the workflow of the graph.

Any alternatives?

eyurtsev commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -1,135 +1,201 @@
+    """Reason for the rejection, if provided."""
+
+
+ActionResponse = ApproveResponse | RejectResponse
```
ActionResponse is also not very self-explanatory. `HumanResponse` or `ReviewResponse` would make more sense

sydney-runkle replied:
Yep, `HumanResponse` is good

sydney-runkle committed:
renaming

sydney-runkle committed:
Merge branch 'sr/hitl-improvements' of https://github.com/langchain-ai/langchain into sr/hitl-improvements
